src/endOfWeek/index.ts

```diff
@@ -32,21 +33,17 @@ import type { LocaleOptions, WeekStartOptions } from '../types'
  */
 export default function endOfWeek(
   dirtyDate: Date | number,
-  dirtyOptions?: LocaleOptions & WeekStartOptions
+  options?: LocaleOptions & WeekStartOptions
 ): Date {
   requiredArgs(1, arguments)
 
-  const options = dirtyOptions || {}
-
-  const locale = options.locale
-  const localeWeekStartsOn =
-    locale && locale.options && locale.options.weekStartsOn
-  const defaultWeekStartsOn =
-    localeWeekStartsOn == null ? 0 : toInteger(localeWeekStartsOn)
-  const weekStartsOn =
-    options.weekStartsOn == null
-      ? defaultWeekStartsOn
-      : toInteger(options.weekStartsOn)
+  const weekStartsOn = toInteger(
+    options?.weekStartsOn ??
+      options?.locale?.options?.weekStartsOn ??
+      _defaultOptions.weekStartsOn ??
+      _defaultOptions.locale?.options?.weekStartsOn ??
+      0
+  )
```

Comment:
I'm comparing this PR to https://github.com/date-fns/date-fns/pull/3087 and this one, with the wider scope, feels more future-prone.

If we decide to move locales to separate packages and make `Intl` the default approach, the change will still make sense. Furthermore, we might think of option autodetection and apply user-specific defaults for `weekStartsOn`, which would be impossible if we stick to locale-only.

It is indeed more code, but we can approach it slightly differently and reduce it to something like:

```suggestion
  const weekStartsOn = getOption(options, 'weekStartsOn')
```

That will also help to abstract the default behavior that we repeat from function to function.